### PR TITLE
Adding yet another Slice<T>

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
There are several competing implementation we are considering for slices.
This one is what we call "three field" implementation.

### CODE REVIEW

**Comment by KrzysztofCwalina** [comment]
@joeduffy

File: src/System.Slices/src/System/Span.cs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+namespace System
+{
+    /// <summary>
+    /// Span is a uniform API for dealing with arrays and subarrays, strings
+    /// and substrings, and unmanaged memory buffers.  It adds minimal overhead
+    /// to regular accesses and is a struct so that creation and subslicing do
+    /// not require additional allocations.  It is type- and memory-safe.
+    /// </summary>
+    public struct Span<T> : IEnumerable<T>
```
Comment (sharwell) [inline src/System.Slices/src/System/Span.cs]: :question: Why not implement `IList<T>`?

File: src/System.Slices/src/System/Span.cs [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        /// <param name="array">The target array.</param>
+        /// <param name="start">The index at which to begin the span.</param>
+        /// <param name="end">The index at which to end the span (exclusive).</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown if the 'array' parameter is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified start or end index is not in range (&lt;0 or &gt;&eq;length).
+        /// </exception>
+        public Span(T[] array, int start, int end)
```
Comment (sharwell) [inline src/System.Slices/src/System/Span.cs]: :exclamation: .NET APIs representing a span are specified by start and count. The constructor needs to follow this pattern. If you also want to provide a way to construct a span from start and end indexes, you can provide a static `Span<T>.FromBounds` helper method.

File: src/System.Slices/src/System/Span.cs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            }
+            else {
+                m_object = null;
+                m_offset = UIntPtr.Zero;
+                Length = 0;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new span over the target unmanaged buffer.  Clearly this
```
Comment (sharwell) [inline src/System.Slices/src/System/Span.cs]: :thought_balloon: I'm not a fan of the way this is worded. I would prefer something like the following:

> Initializes a new instance of the Span&lt;T&gt; struct. This method does not perform bounds checking for the unmanaged memory buffer. The caller should ensure that ptr and length always represent a valid range of memory which can be used as a buffer.

File: src/System.Slices/src/System/Span.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Span.cs]: Maybe at some point we would implement it.

File: src/System.Slices/src/System/Span.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Span.cs]: Yeah, good point. I will fix it.

File: src/System.Slices/src/System/Span.cs (hunk 3 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Span.cs]: BTW, I did not write the original comment (@joeduffy did). But I have to say I like Joe's better. It explains what constitutes a valid buffer (right size for Ts). Your version just says that it needs to be valid.

File: src/System.Buffers/src/System/ByteSpan.cs [hunk 4]
```diff
@@ -76,6 +91,9 @@ public ByteSpan Slice(int index, int count)
 
         public byte[] CreateArray()
         {
+            if(_length == 0) {
+                return new byte[0];
```
Comment (davidfowl) [inline src/System.Buffers/src/System/ByteSpan.cs]: Array.Empty?

Comment (KrzysztofCwalina) [inline src/System.Buffers/src/System/ByteSpan.cs]: Does not exist in System.Runtime 4.0.0

**Pull request merged**